### DISCUSSION TITLE
Initialization of positional encodings?

### DISCUSSION THREAD

**Discussion by btanner** [awf/functional-transformer#4] [General]
Hi there. Great work!

I shared this with a colleague and they were concerned that your example does not seem to initialize the positional encodings beyond zeros.

Should there be a comment or an implementation of setting up the positional encoding?

**Comment by awf** [comment]
Thanks for your interest!

I can see two alternative initializations: sine waves, or random.  Are you thinking about those or another alternative?

In all cases, I think that the general model class is not changed, as the space of explorable embeddings is the same, although obviously in practice the initialization will matter.

I do feel that calling them position encodings is reasonable as they are indexed 0:L, even without any more explicit "position" initialization.

**Comment by btanner** [comment]
**tl;dr** You're learning the positional encodings, which is not super obviously how things are done for a newbie learning about transformers for the first time :)

**Longer**
I think you're exposing my fundamental lack of understanding of transformers, or we're talking a little bit past each other.

...skimming blog posts....

Ok, I think I have made progress. One challenge of an effort like yours is that the people that come to look at your beautiful  and clean code are hoping to understand how transformers work, which means they might not be equipped to navigate nuance and some choices you have made.

If a newbie like me goes spelunking on Google to learn about positional encoding, and whether they are typically hard coded or learned,  I'm going to get the strong impression that positional encodings are fixed and are not learned jointly with the rest of the transformer model. I went looking to find this out and didn't find any suggestion that they were learned. Admittedly I read mostly blog posts on this specific topic, and perhaps I did not read carefully enough.

I walked the code history from your project, to the one you reference:
https://github.com/vpj/jax_transformer/blob/master/transformer.py

I note that in that code they do similar to yours, but within a class called `EmbeddingsWithLearnedPositionalEncoding`. The **learned** part obviously is a hint.

They in turn referenced this code:
https://nn.labml.ai/transformers/models.html#EmbeddingsWithLearnedPositionalEncoding

In that code, they have both `EmbeddingsWithPositionalEncoding` and `EmbeddingsWithLearnedPositionalEncoding`.

So I headed back to Google determined to figure this out, and found this one:
https://www.inovex.de/de/blog/positional-encoding-everything-you-need-to-know/

Halfway down that post I found:

> Another variant of absolute positional encoding exists where the position embeddings are learned jointly with the network model during training.

**Comment by awf** [comment]
Thanks for the elaboration.  I agree, it's not obvious that the embeddings are learned.   Two fixes:
1. Short term -- I'll improve the commments there
2. I'll see how I might include the `sin(t)` terms, because I agree they are how it's initially presented, and it's easier to see their effect than reasoning about the 'learned' ones.

**Reply by awf** [reply]
[Moved from issue to discussion, and made an explicit issue for the sin(t) encodings]

**Reply by btanner** [reply]
I think both of these are great strategies. Thanks for being so responsive!
